db/src/Pos/DB/Block/Slog/Logic.hs

```diff
@@ -145,41 +148,69 @@ slogVerifyBlocks genesisConfig curSlot blocks = runExceptT $ do
     logInfo $ sformat ("slogVerifyBlocks: Consensus era is " % shown) era
     (adoptedBV, adoptedBVD) <- lift getAdoptedBVFull
     let dataMustBeKnown = mustDataBeKnown uc adoptedBV
-    let headEpoch = blocks ^. _Wrapped . _neHead . epochIndexL
-    leaders <- lift $
-        lrcActionOnEpochReason
-            headEpoch
-            (sformat
-                 ("slogVerifyBlocks: there are no leaders for epoch " %build)
-                 headEpoch)
-            LrcDB.getLeadersForEpoch
-    -- We take head here, because blocks are in oldest first order and
-    -- we know that all of them are from the same epoch. So if there
-    -- is a genesis block, it must be head and only head.
-    case blocks ^. _OldestFirst . _neHead of
-        (Left block) ->
-            when (block ^. genBlockLeaders /= leaders) $
-            throwError "Genesis block leaders don't match with LRC-computed"
+    leaders <- case era of
+        Original ->
+            let headEpoch = blocks ^. _Wrapped . _neHead . epochIndexL
+            in lift $
+                lrcActionOnEpochReason
+                    headEpoch
+                    (sformat
+                        ("slogVerifyBlocks Original: there are no leaders for epoch " %build)
+                        headEpoch)
+                    LrcDB.getLeadersForEpoch
+        OBFT ObftStrict -> do
+            initialSlot <- case curSlot of
+                                Just cs -> pure cs
+                                Nothing -> throwError "slogVerifyBlocks ObftStrict: curSlot set to Nothing - \
+                                            \this occurs in EBBs which should not appear"
+            pure $ getEpochSlotLeaderScheduleObft genesisConfig
+                                                  (siEpoch initialSlot)
+        OBFT ObftLenient -> do
+            -- The lenient OBFT block validation algorithm only requires a
+            -- collection of "acceptable" slot leaders rather than a slot
+            -- leader schedule.
+            let gStakeholders = Genesis.configGenesisWStakeholders genesisConfig
```

Comment:
To check my understanding - these are the original genesis StakeholderIds, and haven't been mapped through delegation to the delegatee's. This parallels the [`getEpochSlotLeaderScheduleObft`](https://github.com/input-output-hk/cardano-sl/pull/4059/files#diff-80d48d01b6b4d192800f566e2751f159R166) call above, which also doesn't map through delegation.

Is that because block validation only cares about the original/delegator StakeholderId, not the delegatee? Where do we check that the delegation performed is valid, and the block comes from the correct delegatee?

Reply:
So within [`verifyBlocksPrefix`](https://github.com/input-output-hk/cardano-sl/blob/92a1e976e33a70cf92b11c631c0eb82f60da7b6c/db/src/Pos/DB/Block/Logic/VAR.hs#L74), both `slogVerifyBlocks` and `dlgVerifyBlocks` are called. So delegation checks don't seem to pertain to `slogVerifyBlocks` but is rather the responsibility of `dlgVerifyBlocks`. However, `slogVerifyBlocks`, which ends up calling [`verifyHeader`](https://github.com/input-output-hk/cardano-sl/blob/ec600a7c5622b252b2bd8a1dc053d74f61b60be2/chain/src/Pos/Chain/Block/Logic/Integrity.hs#L97), does perform a leader check in order to make sure that the correct leader has signed the block.

Reply:
To properly answer your question, `dlgVerifyBlocks` cares about verifying the delegatee while `slogVerifyBlocks` doesn't.

Reply:
Now, I'd like to address why `leaders` is either a full round-robin schedule for the current epoch during `ObftStrict` or just a collection of the genesis stakeholders during `ObftLenient`.

This is basically because strict validation adheres to a schedule (as does `Original` validation) while lenient validation _only_ cares whether both:

- a valid leader has minted this block
- and `blocksMintedByLeaderInLastK <= floor $ k * t`

I realize this might be a bit confusing to utilize `leaders` in two different ways depending upon the current `ConsensusEra`.

Reply:
Thanks, I had forgotten about the other `<name>VerifyBlocks` functions.

I'll have to read through those to be sure this all checks out.

RE your last message, that makes sense, and I understood it while reading through from your comments. It's a little confusing, I agree, but seems to add minimal technical debt. The alternative that comes to mind is a datatype which holds the slotleaders for different eras, which we pass instead of `ConsensusEra`, to simplify the case logic. Something like this, perhaps:
```
ConsensusEraLeaders = Original SlotLeaders
                                       | OBFTStrict SlotLeaders
                                       | OBFTLenient (Set StakeholderId)
```
but this seems unnecessarily complex.

Reply:
Ended up deciding to follow your suggestion above.